### PR TITLE
[thrift_proxy] Replace local HeaderMap with Http::HeaderMap[Impl]

### PR DESCRIPTION
**Pull request by brirams**
*Description*:
As part of implementing header matching in the thrift proxy, we'll want to re-use a lot of the existing machinery that exists for http header matching. To make that more straightforward, use the http header implementation in the thrift proxy instead of our own rolled one.

*Risk Level*:
low

*Testing*:
tests, new and old, pass.

### CODE REVIEW

File: test/extensions/filters/network/thrift_proxy/metadata_test.cc [hunk 1]
```diff
@@ -56,8 +56,7 @@ TEST(MessageMetadataTest, Headers) {
   MessageMetadata metadata;
 
   EXPECT_EQ(metadata.headers().size(), 0);
-
-  metadata.addHeader("k", "v");
+  metadata.headers().addReference(Http::LowerCaseString("k"), "v");
```
Comment (zuercher) [inline test/extensions/filters/network/thrift_proxy/metadata_test.cc]: addCopy -- those strings still don't live beyond this line of code.

**Comment by zuercher** [comment]
tag @rgs1 and @fishcakez

**Comment by brirams** [comment]
mistakenly rebased my work vs. merging master. sorry, ya'll.

File: source/extensions/filters/network/thrift_proxy/header_transport_impl.cc [hunk 2]
```diff
@@ -296,7 +300,7 @@ void HeaderTransportImpl::writeVarString(Buffer::Instance& buffer, const std::st
   if (len == 0) {
     return;
   }
-  buffer.add(str);
+  buffer.add(str.data(), str.length());
```
Comment (zuercher) [inline source/extensions/filters/network/thrift_proxy/header_transport_impl.cc]: Just use `len`

File: source/extensions/filters/network/thrift_proxy/header_transport_impl.cc [hunk 3]
```diff
@@ -146,7 +146,7 @@ bool HeaderTransportImpl::decodeFrameStart(Buffer::Instance& buffer, MessageMeta
     while (num_headers-- > 0) {
       std::string key = drainVarString(buffer, header_size, "header key");
```
Comment (rgs1) [inline source/extensions/filters/network/thrift_proxy/header_transport_impl.cc]: nit: couldn't key be constructed here as a `LowerCaseString`?

File: source/extensions/filters/network/thrift_proxy/header_transport_impl.cc [hunk 4]
```diff
@@ -146,7 +146,7 @@ bool HeaderTransportImpl::decodeFrameStart(Buffer::Instance& buffer, MessageMeta
...
       std::string value = drainVarString(buffer, header_size, "header value");
```
Comment (rgs1) [inline source/extensions/filters/network/thrift_proxy/header_transport_impl.cc]: nit: should key/value be consts?

**Pull request merged**